### PR TITLE
Support `CREATE TABLE AS SELECT` and `INSERT` in BigQuery

### LINKED ISSUE #6868: Add INSERT support to BigQuery connector

### LINKED ISSUE #6869: Add CREATE TABLE AS (CTAS) support to BigQuery connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Support `CREATE TABLE AS SELECT` and `INSERT` in BigQuery
Fixes #6868
Fixes #6869

## Documentation

(x) Sufficient documentation is included in this PR.

## Release notes

(x) Release notes entries required with the following suggested text:

```markdown
# BigQuery
* Support `CREATE TABLE AS SELECT` and `INSERT` statements. ({issue}`6868`, {issue}`6869`)
```

### CODE REVIEW

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryInsertTableHandle.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    @JsonCreator
+    public BigQueryInsertTableHandle(
+            @JsonProperty("schemaTableName") SchemaTableName schemaTableName,
+            @JsonProperty("columnNames") List<String> columnNames,
+            @JsonProperty("columnTypes") List<Type> columnTypes)
+    {
+        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+        this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
+        this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+        checkArgument(columnNames.size() == columnTypes.size(), "columnNames and columnTypes should have the same size");
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryInsertTableHandle.java:44]: should -> must

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 2]
```diff
@@ -389,7 +397,16 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         }
     }
 
-    private void createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, Optional<ConnectorTableLayout> layout, RetryMode retryMode)
+    {
+        if (tableMetadata.getComment().isPresent()) {
+            throw new TrinoException(NOT_SUPPORTED, "This connector does not support creating tables with table comment");
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:404]: is it bigquery limitation or we just don't support it yet?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 3]
```diff
@@ -422,6 +452,27 @@ public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle
...
+    public ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> columns, RetryMode retryMode)
+    {
+        if (retryMode != RetryMode.NO_RETRIES) {
+            throw new TrinoException(NOT_SUPPORTED, "This connector does not support query retries");
+        }
+        BigQueryTableHandle table = (BigQueryTableHandle) tableHandle;
+        if (isWildcardTable(TableDefinition.Type.valueOf(table.getType()), table.getRemoteTableName().getTableName())) {
+            throw new TrinoException(BIGQUERY_UNSUPPORTED_OPERATION, "This connector does not support inserting into wildcard tables");
+        }
+        List<String> columnNames = columns.stream().map(column -> ((BigQueryColumnHandle) column).getName()).collect(toImmutableList());
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:465]: JDK 17 has` .toList()` on a `Stream<>` which according to Javadoc returns an unmodifiable list. Should we use it @martint ?

For reference: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html#toList()

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    @JsonCreator
+    public BigQueryOutputTableHandle(
+            @JsonProperty("schemaTableName") SchemaTableName schemaTableName,
+            @JsonProperty("columnNames") List<String> columnNames,
+            @JsonProperty("columnTypes") List<Type> columnTypes)
+    {
+        this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+        this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
+        this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+        checkArgument(columnNames.size() == columnTypes.size(), "columnNames and columnTypes should have the same size");
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java:44]: should -> must

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    private final List<String> columnNames;
+    private final List<Type> columnTypes;
+
+    public BigQueryPageSink(BigQueryClient client, SchemaTableName schemaTableName, List<String> columnNames, List<Type> columnTypes)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(schemaTableName, "schemaTableName is null");
+        this.columnNames = ImmutableList.copyOf(requireNonNull(columnNames, "columnNames is null"));
+        this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
+        checkArgument(columnNames.size() == columnTypes.size(), "columnNames and columnTypes sizes don't match");
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java:71]: `columnNames and columnTypes must have the same size`

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    @Nullable
+    private Object getObjectValue(Type type, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        // TODO: Add support for decimal, time, timestamp, timestamp with time zone, geography, array, map, row type
+        if (type.equals(BooleanType.BOOLEAN)) {
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryPageSink.java:99]: Can we extract this logic to a separate class?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java [hunk 7]
```diff
@@ -298,7 +297,7 @@ private static StandardSQLTypeName toStandardSqlTypeName(Type type)
         if (type == TimestampWithTimeZoneType.TIMESTAMP_TZ_MICROS) {
             return StandardSQLTypeName.TIMESTAMP;
         }
-        if (type instanceof CharType || type instanceof VarcharType) {
+        if (type instanceof VarcharType) {
```
Comment (wendigo) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:300]: why this change?

**Review by wendigo — COMMENTED** [review]
Overall design looks good. We need to improve on type coverage and error handling.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 3 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:465]: unmodifiable list and immutablelist semantics are different. `null` hostility being the big one and unmodifiable isn't really unmodifiable if someone else holds a reference.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 2 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:399]: It's just connector's limitation.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: It was required to fix incorrect predicates between `char` vs `varchar` if I remember correctly.

**Review by wendigo — APPROVED** [review]
lgtm :)

**Comment by ebyhr** [comment]
@hashhar Could you please review this PR when you have time?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 7 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: What is CharType mapped to now? I don't see explicit mapping so does it mean it's unsupported type now?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java [hunk 8]
```diff
@@ -270,6 +273,14 @@ public String selectSql(TableId table, String formattedColumns)
         return format("SELECT %s FROM `%s`", formattedColumns, tableName);
     }
 
+    public void insert(InsertAllRequest insertAllRequest)
+    {
+        InsertAllResponse response = bigQuery.insertAll(insertAllRequest);
+        if (response.hasErrors()) {
+            throw new TrinoException(BIGQUERY_FAILED_TO_EXECUTE_QUERY, format("Failed to insert rows: %s", response.getInsertErrors()));
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:280]: Does BigQuery rollback already inserted rows or leaves table with partial inserts?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 9]
```diff
@@ -393,16 +408,29 @@ private void createTable(ConnectorSession session, ConnectorTableMetadata tableM
             throw new SchemaNotFoundException(schemaName);
         }
 
-        List<Field> fields = tableMetadata.getColumns().stream()
-                .map(column -> toField(column.getName(), column.getType(), column.getComment()))
-                .collect(toImmutableList());
+        ImmutableList.Builder<Field> fields = ImmutableList.builder();
+        ImmutableList.Builder<String> columnsNames = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnsTypes = ImmutableList.builder();
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:413]: nit: since we know the size we can use `builderWithExpectedSize` (or whatever it's called).

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 10]
```diff
@@ -417,6 +445,27 @@ public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle
...
+    public ConnectorInsertTableHandle beginInsert(ConnectorSession session, ConnectorTableHandle tableHandle, List<ColumnHandle> columns, RetryMode retryMode)
+    {
+        if (retryMode != RetryMode.NO_RETRIES) {
+            throw new TrinoException(NOT_SUPPORTED, "This connector does not support query retries");
+        }
+        BigQueryTableHandle table = (BigQueryTableHandle) tableHandle;
+        if (isWildcardTable(TableDefinition.Type.valueOf(table.getType()), table.getRemoteTableName().getTableName())) {
+            throw new TrinoException(BIGQUERY_UNSUPPORTED_OPERATION, "This connector does not support inserting into wildcard tables");
+        }
+        List<String> columnNames = columns.stream().map(column -> ((BigQueryColumnHandle) column).getName()).collect(toImmutableList());
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:458]: With wider tables we'd loop twice. Is it premature to convert to single for loop and use builders to populate `columnTypes` and `columnNames`?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+public class BigQueryOutputTableHandle
+        implements ConnectorOutputTableHandle
+{
+    private final SchemaTableName schemaTableName;
+    private final List<String> columnNames;
+    private final List<Type> columnTypes;
+
+    @JsonCreator
+    public BigQueryOutputTableHandle(
+            @JsonProperty("schemaTableName") SchemaTableName schemaTableName,
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java:37]: Should this be `SchemaTableName` or the remote mapped names?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    private BigQueryTypeUtils() {}
+
+    @Nullable
+    public static Object readNativeValue(Type type, Block block, int position)
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        // TODO: Add support for decimal, time, timestamp, timestamp with time zone, geography, array, map, row type
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java:53]: file issue?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+        if (type.equals(BooleanType.BOOLEAN)) {
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java:54]: static import in this class for all types where possible?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        if (type.equals(VarbinaryType.VARBINARY)) {
+            return Base64.getEncoder().encodeToString(type.getSlice(block, position).getBytes());
+        }
+        if (type.equals(DateType.DATE)) {
+            long days = type.getLong(block, position);
+            return DATE_FORMATTER.format(LocalDate.ofEpochDay(days));
+        }
+
+        throw new TrinoException(NOT_SUPPORTED, "unsupported type: " + type);
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeUtils.java:83]: nit: `unsupported` -> `Unsupported`

**Review by hashhar — COMMENTED** [review]
I still need to look at type mapping tests.

Also we should add tests for inserts into `TestBigQueryCaseInsensitiveMapping` (and also for JDBC connectors but that's separate PR).

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java (hunk 8 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:280]: It doesn't matter since we insert via temp tables but it's useful context to have to make sure future refactorings of direct insert mechanisms don't introduce bugs.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: It's mapped to BigQuery `STRING` type now.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java:281]: It won't be rollbacked. Left a code comment.

**Comment by ebyhr** [comment]
@hashhar Addressed comments.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 7 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: I'm probably being stupid but I can't find code which handles that (mapping Trino char to BigQuery STRING). Can you point me to it?

I'm probably confused because of removal of `char` from `createTableSupportedTypes` in `TestBigQueryConnectorTests`.

**Review by hashhar — APPROVED** [review]
LGTM % some questions.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: Sorry, I misunderstood "now" as "before this change". I wanted to say:
* `CHAR` was mapped to `STRING` before this PR
* `CHAR` is unsupported after this PR

Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryType.java:302]: Sorry for causing confusion. 😄 Do we want to mention this in release notes? The only possible impact I see is some CTAS statements from other catalogs might fail but easy to preserve old "incorrect" behaviour by casting CHAR columns to VARCHAR instead.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java (hunk 11 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryOutputTableHandle.java]: It should be "remote", but a logic to lookup a remote schema name was missing. Could you take a look at https://github.com/trinodb/trino/compare/28038ec2b874963631ae476ca1b9cdcef36e0462..cac3a485decbaebdada4a356c0b50bab68ca34b7?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 15]
```diff
@@ -417,6 +448,32 @@ public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle
...
+            throw new TrinoException(BIGQUERY_UNSUPPORTED_OPERATION, "This connector does not support inserting into wildcard tables");
+        }
+        ImmutableList.Builder<String> columnNames = ImmutableList.builderWithExpectedSize(columns.size());
+        ImmutableList.Builder<Type> columnTypes = ImmutableList.builderWithExpectedSize(columns.size());
+        for (ColumnHandle columnHandle : columns) {
+            BigQueryColumnHandle column = (BigQueryColumnHandle) columnHandle;
+            columnNames.add(column.getName());
+            columnTypes.add(column.getTrinoType());
+        }
+        return new BigQueryInsertTableHandle(table.getRemoteTableName(), columnNames.build(), columnTypes.build());
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:468]: q: It seems following code in BaseJdbcClient is redundant then - https://github.com/trinodb/trino/blob/28a62e1ac5897635cfdf224a434b174a6c274ea4/plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/BaseJdbcClient.java#L619-L620

Because both in BigQuery and BaseJdbcClient we populate the remote names when `getTableHandle` is called.

Or do we need to re-fetch the remote name here?

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 16]
```diff
@@ -468,6 +525,13 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
         return Optional.of(new ConstraintApplicationResult<>(updatedHandle, constraint.getSummary(), false));
     }
 
+    private String getRemoteSchemaName(BigQueryClient client, String projectId, String datasetName)
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:528]: Probably we should also call this in `createSchema` to avoid creating schemas in BigQuery which differ only in case. Separate and pre-existing issue however.

**Review by hashhar — COMMENTED** [review]
Looks good. Thanks.

I'll revisit type mapping and approve.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 15 above)
Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:468]: I think we don't need to refresh here.

Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:468]: That makes sense to me as well. I'll check why JDBC does it though and send a PR accordingly. Thanks.

**Comment by ebyhr** [comment]
Rebased on upstream to resolve conflicts.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 17]
```diff
@@ -2290,6 +2290,8 @@ public void testCreateTableAsSelect()
                         "SELECT 1234567890, 1.23",
                 "SELECT count(*) + 1 FROM customer");
 
+        // TODO: BigQuery throws table not found at BigQueryClient.insert if we reuse the same table name
```
Comment (hashhar) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2293]: Can you explain a bit more to me? Does this mean that when the DROP TABLE above runs and then we try to CREATE TABLE with same name again here it fails?

**Review by hashhar — APPROVED** [review]
LGTM - a question for my understanding.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java (hunk 17 above)
Comment (ebyhr) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2293]: > Does this mean that when the DROP TABLE above runs and then we try to CREATE TABLE with same name again here it fails?

Yes. Also, the issue doesn't happen in case of empty tables. It seems there's cache or something for `BigQuery.insertAll`. I couldn't find relevant options in the library at glance. I will look into the detail later.

**Review by hashhar — APPROVED** [review]
LGTM. Thanks for working on this.

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java [hunk 18]
```diff
@@ -408,7 +415,13 @@ public void createTable(ConnectorSession session, ConnectorTableMetadata tableMe
         }
     }
 
-    private void createTable(ConnectorSession session, ConnectorTableMetadata tableMetadata)
+    @Override
+    public ConnectorOutputTableHandle beginCreateTable(ConnectorSession session, ConnectorTableMetadata tableMetadata, Optional<ConnectorTableLayout> layout, RetryMode retryMode)
+    {
+        return createTable(session, tableMetadata);
```
Comment (hashhar) [inline plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:421]: Does this also need a defensive check to verify query retries are not enabled?

Comment (ebyhr) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:421]: I think so. Thanks for catching that. Do you want to send the PR?

Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:421]: https://github.com/trinodb/trino/pull/13812

File: plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java (hunk 16 above)
Comment (hashhar) [reply plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java:571]: Turns out this is not needed/the wrong thing to do. When creating schemas we already list them and the listing will include the lowercase name so Trino will see a schema already exists - I've added a test to verify this similar to what you added for DROP SCHEMA in https://github.com/trinodb/trino/pull/13812

File: docs/src/main/sphinx/connector/bigquery.rst [hunk 19]
```diff
@@ -208,12 +208,14 @@ SQL support
 -----------
 
 The connector provides read and write access to data and metadata in the
-BigQuery database, though write access is limited. In addition to the
```
Comment (colebow) [inline docs/src/main/sphinx/connector/bigquery.rst:211]: 😃
